templates/hooks--fsmonitor-watchman.sample

```diff
@@ -3,107 +3,141 @@
 use strict;
 use warnings;
 use IPC::Open2;
+use File::Spec;
+use File::Basename;
 
 # An example hook script to integrate Watchman
 # (https://facebook.github.io/watchman/) with git to speed up detecting
 # new and modified files.
 #
-# The hook is passed a version (currently 1) and a time in nanoseconds
-# formatted as a string and outputs to stdout all files that have been
-# modified since the given time. Paths must be relative to the root of
-# the working tree and separated by a single NUL.
+# The hook is passed a version (currently 2) and last update token
+# formatted as a string and outputs to stdout a new update token and
+# all files that have been modified since the update token. Paths must
+# be relative to the root of the working tree and separated by a single NUL.
 #
 # To enable this hook, rename this file to "query-watchman" and set
 # 'git config core.fsmonitor .git/hooks/query-watchman'
 #
-my ($version, $time) = @ARGV;
+my ($version, $last_update_token) = @ARGV;
 
 # Check the hook interface version
 
-if ($version == 1) {
-	# convert nanoseconds to seconds
-	# subtract one second to make sure watchman will return all changes
-	$time = int ($time / 1000000000) - 1;
-} else {
+if ($version ne 2) {
 	die "Unsupported query-fsmonitor hook version '$version'.\n" .
 	    "Falling back to scanning...\n";
 }
 
-my $git_work_tree;
-if ($^O =~ 'msys' || $^O =~ 'cygwin') {
-	$git_work_tree = Win32::GetCwd();
-	$git_work_tree =~ tr/\\/\//;
-} else {
-	require Cwd;
-	$git_work_tree = Cwd::cwd();
-}
```

Comment:
@dscho I wasn't sure if the `Cwd` would always be the root of the repo or might be somewhere else in the repository that is why and added the `get_working_dir()` to make sure we are in the root before we add the directory to the watch list.

Reply:
I _think_ that the hooks are always called from the top-level directory. If in doubt, maybe there is an environment variable `GIT_WORK_TREE` that you can use?

Reply:
This [was discussed recently on-list](https://lore.kernel.org/git/pull.466.git.1574374826.gitgitgadget@gmail.com/T/#t). It looks like `GIT_DIR` is only set if it differs from normal, but maybe `GIT_WORK_TREE` has similar characteristics? At the least, `t3030-merge-recursive.sh` has some `GIT_WORK_TREE` strangeness that can help.